tests/testthat/test-yoda_test_linter.R

```diff
@@ -0,0 +1,44 @@
+test_that("yoda_test_linter skips allowed usages", {
+  expect_lint("expect_equal(x, 2)", NULL, yoda_test_linter())
+  # namespace qualification doesn't matter
+  expect_lint("testthat::expect_identical(x, 'a')", NULL, yoda_test_linter())
+  # two variables can't be distinguished which is expected/actual (without
+  #   playing quixotic games trying to parse that out from variable names)
+  expect_lint("expect_equal(x, y)", NULL, yoda_test_linter())
+})
+
+test_that("yoda_test_linter blocks simple disallowed usages", {
+  expect_lint(
+    "expect_equal(2, x)",
+    rex::rex("Tests should compare objects in the order 'actual', 'expected'"),
+    yoda_test_linter()
+  )
+  expect_lint(
+    "testthat::expect_identical('a', x)",
+    rex::rex("Tests should compare objects in the order 'actual', 'expected'"),
+    yoda_test_linter()
+  )
+  expect_lint(
+    "expect_setequal(2, x)",
+    rex::rex("Tests should compare objects in the order 'actual', 'expected'"),
+    yoda_test_linter()
+  )
+  # complex literals are slightly odd
+  expect_lint(
+    "expect_equal(2 + 1i, x)",
+    rex::rex("Tests should compare objects in the order 'actual', 'expected'"),
+    yoda_test_linter()
+  )
+})
+
+# TODO(michaelchirico): Should this be extended to RUnit tests? It seems yes,
+#   but the argument names in RUnit (inherited from base all.equal()) are a bit
+#   confusing, e.g. `checkEqual(target=, current=)`. From the name, one might
+#   reasonably conclude 'expected' comes first, and 'actual' comes second.
+# TODO(michaelchirico): What sorts of combinations of literals can be included?
+#   e.g. expect_equal(c(1, 2), x) is a yoda test; is expect_equal(c(x, 1), y)?
+#   clearly it's not true for general f() besides c(). What about other
+#   constructors of literals? data.frame(), data.table(), tibble(), ...?
```

Comment:
Maybe it's worth to craft up an "extended literal" XPath that contains a few whitelisted functions and use that throughout lintr if we want to match a "constant" expression?

First thought would be something like `not(SYMBOL_FUNCTION_CALL[not(text() = 'c' or ...)])` etc.

Reply:
sounds great, I was thinking something similar. We have something like that in `StringsAsFactorsLinter` as well (which also has some logic about mixing in `rep()`), and it can be re-used in the `expect_equal($LITERAL, typeof(x))` family of lints too.

Reply:
Let's mark this as a TODO for now. we have a messy XPath that works for a limited number of cases, I'd like to generalize it if possible.

For reference here's the XPath for `StringsAsFactorsLinter` (so we only look for `STR_CONST`):

```
//expr[
  expr[SYMBOL_FUNCTION_CALL[text() = 'data.frame']]
  and expr[
    (
      STR_CONST
      or (
        expr[SYMBOL_FUNCTION_CALL[text() = 'c']]
        and expr[STR_CONST]
        and not(expr[SYMBOL or expr])
      )
      or expr[
        SYMBOL_FUNCTION_CALL[text() = 'rep']
        and (
          following-sibling::expr[1][
            STR_CONST
            or (
              expr[SYMBOL_FUNCTION_CALL[text() = 'c']]
              and expr[STR_CONST]
              and not(expr[SYMBOL or expr])
            )
          ]
        )
      ] or expr[SYMBOL_FUNCTION_CALL[
        text() = 'character'
        or text() = 'as.character'
        or text() = 'paste'
        or text() = 'sprintf'
        or text() = 'format'
        or text() = 'formatC'
        or text() = 'prettyNum'
        or text() = 'toString'
        or text() = 'encodeString'
      ]]
    )
    and not(preceding-sibling::SYMBOL_SUB[1][text() = 'row.names'])
  ]
  and not(SYMBOL_SUB[text() = 'stringsAsFactors'])
]
```